# GitLab NGINX Patch: Support templating LoadBalancer IP to support multiple
# NGINX charts with different external IPs for GitLab Geo deployments.
diff --git a/charts/nginx-ingress/templates/_helpers.tpl b/charts/nginx-ingress/templates/_helpers.tpl
index 6cbda2d4d..a1e4ecffa 100644
--- a/charts/nginx-ingress/templates/_helpers.tpl
+++ b/charts/nginx-ingress/templates/_helpers.tpl
@@ -261,3 +261,27 @@ Extra modules.
     - name: modules
       mountPath: /modules_mount
 {{- end -}}
+
+{{/*
+  Returns the load balancer IP sourced .Values.externalIpTPl.
+  By default this template renders the value of .Values.global.hosts.externalIP.
+*/}}
+{{- define "ingress-nginx.controller.service.globalLoadBalancerIP" -}}
+{{ tpl .Values.externalIpTpl . }}
+{{- end -}}
+
+{{/*
+Returns the load balancer IP for the controller service.
+
+Uses the value rendered from `externalIpTPl` and falls back
+to the .Values.controller.service.loadBalancerIP.
+Ensures that the Geo Ingress Controller does not attempt to
+use the same IP address as the main ingress.
+*/}}
+{{- define "ingress-nginx.controller.service.loadBalancerIP" -}}
+{{- $globalLbIp := include "ingress-nginx.controller.service.globalLoadBalancerIP" . -}}
+{{- $lbIp := coalesce $globalLbIp .Values.controller.service.loadBalancerIP -}}
+{{- if $lbIp -}}
+loadBalancerIP: {{ $lbIp }}
+{{- end -}}
+{{- end -}}
diff --git a/charts/nginx-ingress/templates/controller-service.yaml b/charts/nginx-ingress/templates/controller-service.yaml
index cb78a7035..e082c958d 100644
--- a/charts/nginx-ingress/templates/controller-service.yaml
+++ b/charts/nginx-ingress/templates/controller-service.yaml
@@ -22,9 +22,7 @@ spec:
 {{- if .Values.controller.service.externalIPs }}
   externalIPs: {{ toYaml .Values.controller.service.externalIPs | nindent 4 }}
 {{- end }}
-{{- if .Values.controller.service.loadBalancerIP }}
-  loadBalancerIP: {{ .Values.controller.service.loadBalancerIP }}
-{{- end }}
+{{- include "ingress-nginx.controller.service.loadBalancerIP" . | nindent 2 }}
 {{- if .Values.controller.service.loadBalancerSourceRanges }}
   loadBalancerSourceRanges: {{ toYaml .Values.controller.service.loadBalancerSourceRanges | nindent 4 }}
 {{- end }}
diff --git a/charts/nginx-ingress/values.yaml b/charts/nginx-ingress/values.yaml
index c58c0c791..76ace60b3 100644
--- a/charts/nginx-ingress/values.yaml
+++ b/charts/nginx-ingress/values.yaml
@@ -1202,3 +1202,5 @@ dhParam: ""
 # Ref: https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx/examples/tcp
 ##
 tcpExternalConfig: ""
+
+externalIpTpl: '{{ .Values.global.hosts.externalIP }}'
